### PR TITLE
snippetexpander: init at 1.0.1

### PR DESCRIPTION
**Pull request by ianmjones**
## Description of changes

Add Snippet Expander daemon, auto expander daemon, GUI, and CLI.

Website: https://snippetexpander.org
Source: https://git.sr.ht/~ianmjones/snippetexpander/

The daemon is a required dependency of the auto expander daemon, GUI and CLI, hence why they are all being submitted in this initial PR together.

It will then be possible for people to choose whether they want to use the GUI, CLI, or auto expander daemon packages in any combination, with the daemon always being included.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
> It will then be possible for people to choose whether they want to use the GUI, CLI, or auto expander daemon packages in any combination, with the daemon always being included.

It would be nicer if this would be one package with different outputs but due to different go.mod files which is rather unusual, this doesn't work.

File: pkgs/by-name/sn/snippetexpanderd/package.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+buildGoModule rec {
+  pname = "snippetexpanderd";
+  version = "1.0.1";
+
+  src = fetchFromSourcehut {
+    owner = "~ianmjones";
+    repo = "snippetexpander";
+    rev = "v${version}";
+    hash = "sha256-y3TJ+L3kXYfZFzAD1vmhvP6Yarctu5LHq/74005h8sI=";
+  };
```
[the remark this answers is not in the document]
Comment (ianmjones) [reply pkgs/by-name/sn/snippetexpanderd/package.nix:22]: How would we do that?

Would I create an extra file in snippetexpanderd's directory with just the `src` expression and then require it in each of packages?

Would I have to move the `version` too?

**Comment by ianmjones** [comment]
> It would be nicer if this would be one package with different outputs but due to different go.mod files which is rather unusual, this doesn't work.

Hmm, I wonder, if I didn't use `buildGoModule`, and created a single `snippetexpander` package with the dependencies for all of the binaries in it and let `make` do its thing, would that open it up to using multiple outputs?

File: pkgs/by-name/sn/snippetexpander/package.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    owner = "~ianmjones";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-y3TJ+L3kXYfZFzAD1vmhvP6Yarctu5LHq/74005h8sI=";
+  };
+
+  vendorHash = "sha256-wSAho59yxcXTu1zQ5x783HT4gtfSM4GdsOEeC1wfHhE=";
+
+  proxyVendor = true;
+
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/by-name/sn/snippetexpander/package.nix:23]: We usually only set it when paths don't work on case insensitive file systems but if you are sure fine for me

File: pkgs/by-name/sn/snippetexpanderd/package.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/by-name/sn/snippetexpanderd/package.nix:22]: I think the easiest is to just inherit version and src from the main package.

**Comment by SuperSandro2000** [comment]
> Hmm, I wonder, if I didn't use `buildGoModule`, and created a single `snippetexpander` package with the dependencies for all of the binaries in it and let `make` do its thing, would that open it up to using multiple outputs?

You want to use buildGoModule because otherwise you need to create the fod and vendor directory manually which is a pain.

I think upstream should just share the go.mod's for all packages and build the different binaries from different subdirs. See https://github.com/sapcc/gatekeeper-addons/ for an example of that.

**Comment by ianmjones** [comment]
@SuperSandro2000 thank you for your advice regarding sharing the `src` and `version`, I've now set things up so that the packages that depend on `snippetexpanderd` now inherit its `src` and `version` as the dependency is already there due to it being required for any of the others to be useful.

I've also retested which packages need `proxyVendor` and found that `snippetexpanderd` doesn't need it, but the other two go packages do.

Prompted by your comment on the weirdness of the project's multiple go.mod files, I've taken another look at it and the next version will be using a single top level go.mod, although for a bunch of reasons I'll be retaining the multiple Makefile setup. Thanks for bringing this weirdness to my attention, much appreciated.

Thanks for your review so far, everything is updated and retested locally by myself, please could you take another look for me?

File: pkgs/by-name/sn/snippetexpandergui/package.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    make man
+    installManPage snippetexpandergui.1
+  '';
+
+  postFixup = ''
+    wrapProgram $out/bin/snippetexpandergui \
+      --prefix XDG_DATA_DIRS : ${gsettings-desktop-schemas}/share/gsettings-schemas/${gsettings-desktop-schemas.name}:${gtk3}/share/gsettings-schemas/${gtk3.name}
+  '';
+
+  doCheck = false;
```
Comment (pbsds) [inline pkgs/by-name/sn/snippetexpandergui/package.nix:62]: Please comment why the checks are disabled.

File: pkgs/by-name/sn/snippetexpanderx/package.nix [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    snippetexpanderd
+  ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    install -m555 snippetexpanderx $out/bin/
+    installManPage snippetexpanderx.1
+  '';
+
+  doCheck = false;
```
Comment (pbsds) [inline pkgs/by-name/sn/snippetexpanderx/package.nix:51]: here too

File: pkgs/by-name/sn/snippetexpanderx/package.nix [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    libgee
+    xorg.libX11
+    snippetexpanderd
+  ];
+
+  installPhase = ''
+    mkdir -p $out/bin
+    install -m555 snippetexpanderx $out/bin/
+    installManPage snippetexpanderx.1
+  '';
```
Comment (pbsds) [inline pkgs/by-name/sn/snippetexpanderx/package.nix:49]:
```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin
    install -m555 snippetexpanderx $out/bin/
    installManPage snippetexpanderx.1
    runHook postInstall
  '';
```

**Review by pbsds — COMMENTED** [review]
- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [ ] executables tested on x86_64-linux
    * webkitgtk is not in cache yet, will test after fixes
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [ ] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

File: pkgs/by-name/sn/snippetexpandergui/package.nix (hunk 3 above)
Comment (ianmjones) [reply pkgs/by-name/sn/snippetexpandergui/package.nix:62]: This is probably a mistake as it does have tests that can be run, I'll remove this line.

File: pkgs/by-name/sn/snippetexpanderx/package.nix (hunk 5 above)
Comment (ianmjones) [reply pkgs/by-name/sn/snippetexpanderx/package.nix:49]: Great suggestion, I'll add these hooks in.

File: pkgs/by-name/sn/snippetexpanderx/package.nix (hunk 4 above)
Comment (ianmjones) [reply pkgs/by-name/sn/snippetexpanderx/package.nix:51]: There aren't any tests in the source for this package, so I think I'm supposed to add this to save a little bit of time?

**Comment by ianmjones** [comment]
Thanks @pbsds, all fixed up.

File: pkgs/by-name/sn/snippetexpanderx/package.nix (hunk 4 above)
Comment (pbsds) [reply pkgs/by-name/sn/snippetexpanderx/package.nix:53]:
```suggestion
  # There are no tests 
  doCheck = false;
```

**Comment by pbsds** [comment]
just [a final nit](https://github.com/NixOS/nixpkgs/pull/269750#discussion_r1429191390)

**Comment by ianmjones** [comment]
> just [a final nit](https://github.com/NixOS/nixpkgs/pull/269750#discussion_r1429191390)

Nit picked! :smile:

**Pull request merged**